reports-backend/src/main/java/org/jboss/da/listings/impl/service/ArtifactServiceImpl.java

```diff
@@ -35,10 +43,22 @@ public boolean isArtifactPresent(GAV gav) {
 
     @Override
     public boolean removeArtifact(String groupId, String artifactId, String version) {
-        T artifact = getDAO().findArtifact(groupId, artifactId, version);
-        if (artifact != null) {
-            getDAO().delete(artifact);
-            return true;
+        Optional<T> artifact = getDAO().findArtifact(groupId, artifactId, version);
+        if (artifact.isPresent()) {
+            if (artifact.get() instanceof BlackArtifact) {
+                getDAO().delete(artifact.get());
+                return true;
+            }
+            if (artifact.get() instanceof WhiteArtifact) {
+                List<Product> products = productDAO.findProductOfArtifact(groupId, artifactId,
+                        version);
+                for (Product p : products) {
+                    p.removeArtifact((WhiteArtifact) artifact.get());
+                    productDAO.update(p);
+                }
+                getDAO().delete(artifact.get());
+                return true;
+            }
```

Comment:
Did you think about implementing this method in BlackArtifactServiceImpl and WhiteArtifactServiceImpl? I think this is typical case of polymorphism.